### PR TITLE
KAFKA-6906: Fixed to commit transactions if data is produced via wall clock punctuation

### PR DESCRIPTION
**Pull request by jadireddi**
https://issues.apache.org/jira/browse/KAFKA-6906

Fixed `StreamTask` to commit transactions if the data is produced via wall-clock punctuation too.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jadireddi** [comment]
@mjsax , Could you please review and let me know if any changes needed.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (jadireddi) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: @mjsax , By removing `commitOffsetNeeded`, StreamThreadTest#shouldCloseTaskAsZombieAndRemoveFromActiveTasksIfProducerWasFencedWhileProcessing got broken. 
 The commit is been happening  at this line, when 1st `thread.runOnce(-1);` from the above `@test`  method getting called.
Earlier this execution is being stopped by `commitOffsetNeeded` . Need to fix this.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Do we need `transactionInFlight` here?  Previously we just considered if `eosEnabled` for committing a transaction.

**Review by bbejeck — COMMENTED** [review]
Thanks for the PR @jadireddi.  I have just one minor comment, but otherwise this LTGM as per the Jira ticket.  But I'll defer to @mjsax and @guozhangwang for their reviews.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Wondering the same. Is EOS is enabled, I would assume that `transactionInFlight` is always `true` when we hit this point?

Comment (jadireddi) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: @bbejeck @mjsax . This fixed is aimed at test case `shouldCloseProducerOnCloseWhenEosEnabled`.
When we are trying to close `statelesstask` is with parameters `clean:true` and `isZonble:false`, 
1) It goes to `StreamTask#commitOffsets` and if `transactionInFlight ` flag is not present in condition check , it is calling `producer.sendOffsetsToTransaction(consumedOffsetsAndMetadata, applicationId);`
2) And call goes to `MockProducer#sendOffsetsToTransaction` . And it verifies for `      verifyNoTransactionInFlight();`. As `transactionInFlight =false`, it throwing `There is no open transaction.` Exception.
3) So included `transactionInFlight ==true` check to filter stateless tasks with clean close.
4) Earlier this piece was not not called due to `commitOffsetNeeded ==false` for this case.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: For `shouldCloseProducerOnCloseWhenEosEnabled`, we could update it by adding `task.initializeTopology()` before `task.close()`. The former would call `producer.beginTransaction` which will then fix this issue.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: This is not a feedback: as we are changing the main loop of StreamThread, we may need to carefully benchmark if this change along with the main loop changes will have unexpected performance penalty: with low traffic input stream, we are effectively sending sync commit requests more frequently. cc @mjsax 

If it does become a problem for performance, we could consider making the commit request async, and consider a commit only completed after the commit response is returned. Of course it means more complicated logic.

**Review by guozhangwang — COMMENTED** [review]
Could we add a unit test for this bug, and validate that without the change this unit test will fail?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (jadireddi) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Fixed

**Comment by jadireddi** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Ack.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Thinking about this once more, we could actually keep the `commitOffsetNeeded` flag, and instead extract 
```
if (eosEnabled) {
    producer.commitTransaction();
    transactionInFlight = false;
    if (startNewTransaction) {
        producer.beginTransaction();
        transactionInFlight = true;
    }
}
```
If we put this after of the `if (commitOffsetNeeded) { }` block, we should be fine.
Note, that within the block, we still need:
```
if (eosEnabled) {
    producer.sendOffsetsToTransaction(consumedOffsetsAndMetadata, applicationId);
} else {
    consumer.commitSync(consumedOffsetsAndMetadata);
}
```

@guozhangwang WDYT about this?

**Review by mjsax — COMMENTED** [review]
One more comment: I think we should add a test that emit data from a punctuation call with EOS enabled.

**Comment by mjsax** [comment]
@jadireddi Code freeze is close and thus I wanted to follow up on this. Do you plan to add a test? If we get this merged, we can also merge your other PR that is blocked by this one.

**Comment by jadireddi** [comment]
@mjsax , For the test method, can we merge this code and rebase blocked ticket pull(https://github.com/apache/kafka/pull/4912) to latest , remove the work around done for passing the `testTopologyTestDriverTest#shouldPunctuateIfWallClockTimeAdvances when eosEnabled is true`
and see `shouldPunctuateIfWallClockTimeAdvances` getting passed. I think this acts as proof for the 
 wall-clock punctuation working. 
I already tested in local. It works.. Kindly let me know your inputs.

**Comment by mjsax** [comment]
I would prefer an explicit test, because for `shouldPunctuateIfWallClockTimeAdvances` (with EOS) it's just a "side effect". If we change the `shouldPunctuateIfWallClockTimeAdvances` test for whatever reason, we would loose test coverage without noticing it. Does this make sense?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Even without EOS and assuming KIP-211 is in place, we'd probably still to commit regularly as a pre-requisite to fix https://issues.apache.org/jira/browse/KAFKA-6502.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: EDIT: thinking about it again, I agree that we could separate them into two issues before and after the 2.0 deadline: in this PR that is going into 2.0 we can still keep the flag and do what @mjsax suggested. And the general refactoring to remove the flag can be delayed to post 2.0.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Fixing KAFKA-6502 might also not require to remove the flag -- keeping both separate makes sense to me.

@jadireddi Can you please address this comment and add a test. Thanks!

**Comment by jadireddi** [comment]
@mjsax , It's true.Agreed. Can i include both +ve and -ve test cases for `wall-clock punctuation`  in `testTopologyTestDriverTest` ?

**Comment by mjsax** [comment]
Because this change affects `StreamTask`, the unit test should be part of `StreamTaskTest` IMHO. Something like: `shouldAlwaysCommitIfEosEnabled`

**Comment by jadireddi** [comment]
@mjsax , Can you please let me know, the below test case have all the test coverage steps or  am i deviating?.

```
    @Test
    public void shouldAlwaysCommitIfEosEnabled() {

        task = createStatelessTask(createConfig(true));
        task.initializeStateStores();
        task.initializeTopology();
        task.punctuate(processorSystemTime, 5, PunctuationType.WALL_CLOCK_TIME, new Punctuator() {
            @Override
            public void punctuate(final long timestamp) {
                task.context().forward(2, 4);
            }
        });
        task.commit();
        assertEquals(producer.history().size() , 1); //Actual value(1): still experimental.

    }
```

**Comment by jadireddi** [comment]
If the above test case is valid . Need to write `MockSinkNode`(which extends SinkNode) class. The reason is :
1) Assume, adding one `MockSinkNode` object as clild to `MockProcessorNode` in `createStatelessTask`
2)  `task.context().forward(2, 4);` calls `ProcessorContextImpl#forward(K, V, processor.To)` and at step 1 as one child got added,  if (children.size() == 1) , block gets executed.
3)   `child.process(key, value);` gets invoked, and it calls `MockSinkNode#Process()`. And `RecordCollectorImpl#send()` gets called . Later call goes to `MockProducer#send(ProducerRecord<K,V>, Callback)` and record gets added t0 `uncommittedSends`
4) During task.commit(), `uncommittedSends` added to `sent` list. And Producer.history.size() gives necessary value.

Can you please advise, if i am deviating.

**Comment by mjsax** [comment]
Sounds reasonable to me.

**Comment by jadireddi** [comment]
@mjsax , Can you please review test case  emitting data from a punctuation call with EOS enabled. Currently calling `recordCollector.send()` directly from punctuate instead of creating `MockSinkNode` and calling.

**Comment by jadireddi** [comment]
retest this please

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: if if should throw here, the lines below can be removed

Also, we should not use `expected` annotation because if `createStatelessTask()` or `createConfig()` throw, the test should fail but not pass.

rewrite to:
```
try {
    task.close(true, false);
    fail("should have throw IllegalStateException");
} catch (final IllegalStateException expected) {
    // pass
}
```

**Review by mjsax — COMMENTED** [review]
Test looks good overall. Some minor comments. Also, please address this comment: https://github.com/apache/kafka/pull/5105/files#r192507791

Thanks a lot!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (jadireddi) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Fixed

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (jadireddi) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: Fixed

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (jadireddi) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Fixed.

**Review by mjsax — APPROVED** [review]
LGTM.

**Comment by jadireddi** [comment]
This ticket made me to learn lot of things. Thank you All for the support and accepting PR.

**Comment by mjsax** [comment]
Thanks for the PR. Glad you learned something :)

Merged to `trunk` and cherry-picked to `2.0`, and `1.1` (hit some issues for `1.0` branch so I did not back port it further)

**Comment by mjsax** [comment]
Cherry-picked to `1.0` and `0.11.0` branches.

**Pull request merged**
